### PR TITLE
recording vs sampled explained

### LINKED ISSUE #223: Clarify relationship between IsSampled and IsRecordingEvents
The relationship between `IsSampled` and `IsRecordingEvents` is not clear. For instance span builder in java allows providing sampler but also setting `setRecordingEvents`. What is the relationship between these two? Which API should be used when? 

* `IsRecordingEvents` - https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#isrecordingevents
samping
 * `IsSampled` - https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-sampling.md#issampled

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
Fixes #223

### CODE REVIEW

File: specification/api-tracing.md [hunk 1]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
+Returns true if this `Span` is recording information like events with the
+`AddEvent` operation, attributes using `SetAttributes`, status with `SetStatus`,
+etc.
 
 There should be no parameter.
 
-Returns true if this `Span` is active and recording information like events with
-the `AddEvent` operation and attributes using `SetAttributes`.
+This flag SHOULD be used to avoid expensive computations of a Span attributes or
+events in case when Span is definitely will not be recorded. Note, child spans
```
Comment (z1c0) [inline specification/api-tracing.md]: "Broken" sentence: is ... will

File: specification/api-tracing.md [hunk 2]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
+recording is determined independently from the value of this flag. Typically
+based on `sampled` bit of a `TraceFlag` from [SpanContext](#spancontext).
+
+This flag value may be `true` despite the entire trace is sampled out. This
```
Comment (Oberon00) [inline specification/api-tracing.md]:
```suggestion
This flag may be `true` despite the entire trace being sampled out. This
```

File: specification/api-tracing.md [hunk 3]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
+allows to record and process information about the individual Span without
+sending it to the backend. The example of this scenario may be recording and
+processing of all incoming requests for the processing and building of an
+SLA/SLO latency charts while sending only subset - sampled spans - to the
+backend. See more in [sampling section of SDK design](sdk-tracing.md#sampling).
+
```
Comment (z1c0) [inline specification/api-tracing.md]: typo: IsRecord**ing**Event

File: specification/api-tracing.md [hunk 4]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
-Returns true if this `Span` is active and recording information like events with
-the `AddEvent` operation and attributes using `SetAttributes`.
+This flag SHOULD be used to avoid expensive computations of a Span attributes or
+events in case when Span is definitely will not be recorded. Note, child spans
+recording is determined independently from the value of this flag. Typically
+based on `sampled` bit of a `TraceFlag` from [SpanContext](#spancontext).
+
+This flag value may be `true` despite the entire trace is sampled out. This
+allows to record and process information about the individual Span without
+sending it to the backend. The example of this scenario may be recording and
```
Comment (Oberon00) [inline specification/api-tracing.md]:
```suggestion
sending it to the backend. An example of this scenario may be recording and
```

File: specification/api-tracing.md [hunk 5]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
+processing of all incoming requests for the processing and building of an
```
Comment (Oberon00) [inline specification/api-tracing.md]:
```suggestion
processing of all incoming requests for the processing and building of
```

File: specification/api-tracing.md [hunk 6]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
+SLA/SLO latency charts while sending only subset - sampled spans - to the
```
Comment (Oberon00) [inline specification/api-tracing.md]:
```suggestion
SLA/SLO latency charts while sending only a subset - sampled spans - to the
```

File: specification/sdk-tracing.md [hunk 7]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+    status, etc.), otherwise all tracing events are dropped.
+  * Users can use this property to determine if expensive trace events can be
+    avoided.
+  * [Span Processor](#span-processor) will receive all spans with this flag set.
+    However, [Span Exporter](#span-exporter) will not receive them unless the
+    `Sampled` flag was set.
+* `Sampled` `TraceFlag` on `SpanContext`
+  * This flag is propagated via the `TraceFlag` to the child Spans. For more
+    details see the w3c definition [here][trace-flags].
+  * This flag indicates that `Span` is being `sampled` and exported.
```
Comment (z1c0) [inline specification/sdk-tracing.md]: the/a Span

File: specification/api-tracing.md (hunk 6 above)
Comment (Oberon00) [inline specification/api-tracing.md]: The ` - `-part is unclear here: It the sent subset identical to the set of all sampled spans or is it a subset of all sampled spans?

File: specification/sdk-tracing.md [hunk 8]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
 All other sampling algorithms may be implemented on SDK layer in exporters, or
 even out of process in Agent or Collector.
 
-API defines two interfaces - [`Sampler`](#sampler) and [`Decision`](#decision)
+OpenTelemetry API has two properties responsible for the data collection:
+
+* `IsRecordEvents` field of a `Span`
```
Comment (z1c0) [inline specification/sdk-tracing.md]: IsRecord**ing**Events

File: specification/api-tracing.md [hunk 9]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
+backend. See more in [sampling section of SDK design](sdk-tracing.md#sampling).
```
Comment (Oberon00) [inline specification/api-tracing.md]:
```suggestion
backend. See also the [sampling section of SDK design](sdk-tracing.md#sampling).
```

File: specification/sdk-tracing.md (hunk 7 above)
Comment (z1c0) [inline specification/sdk-tracing.md]: the/a Span

File: specification/api-tracing.md [hunk 10]
```diff
@@ -287,12 +287,27 @@ The Span interface MUST provide:
...
+
+Users of the API should only access the `IsRecordEvents` property when
```
Comment (Oberon00) [inline specification/api-tracing.md]:
```suggestion
Users of the API should only access the `IsRecording` property when
```

File: specification/sdk-tracing.md [hunk 11]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
 All other sampling algorithms may be implemented on SDK layer in exporters, or
 even out of process in Agent or Collector.
 
-API defines two interfaces - [`Sampler`](#sampler) and [`Decision`](#decision)
+OpenTelemetry API has two properties responsible for the data collection:
```
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
The OpenTelemetry API has two properties responsible for the data collection:
```

File: specification/sdk-tracing.md [hunk 12]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+    However, [Span Exporter](#span-exporter) will not receive them unless the
+    `Sampled` flag was set.
+* `Sampled` `TraceFlag` on `SpanContext`
+  * This flag is propagated via the `TraceFlag` to the child Spans. For more
+    details see the w3c definition [here][trace-flags].
+  * This flag indicates that `Span` is being `sampled` and exported.
+  * [Span Processor](#span-processor) and [Span Exporter](#span-exporter) will
+    receive spans with the `Sampled` flag set for processing.
+
+The flag combination `SampledFlag == false` and `IsRecordEvents == true` means
```
Comment (z1c0) [inline specification/sdk-tracing.md]: Either `RecordEvents` or `IsRecordingEvents`?

File: specification/sdk-tracing.md (hunk 8 above)
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
* `IsRecording` field of a `Span`
```

File: specification/sdk-tracing.md [hunk 13]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+that the current `Span` does record information, but most likely the child
+`Span` will not.
+
+The flag combination `SampledFlag == true` and `RecordEvents == false` can cause
```
Comment (z1c0) [inline specification/sdk-tracing.md]: Either `RecordEvents` or `IsRecordingEvents`?

File: specification/sdk-tracing.md [hunk 14]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
 
-API defines two interfaces - [`Sampler`](#sampler) and [`Decision`](#decision)
+OpenTelemetry API has two properties responsible for the data collection:
+
+* `IsRecordEvents` field of a `Span`
+  * If `true` the current `Span` records tracing events (attributes, events,
+    status, etc.), otherwise all tracing events are dropped.
+  * Users can use this property to determine if expensive trace events can be
+    avoided.
+  * [Span Processor](#span-processor) will receive all spans with this flag set.
```
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
  * [Span Processors](#span-processor) will receive all spans with this flag set.
```

File: specification/sdk-tracing.md [hunk 15]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
 All other sampling algorithms may be implemented on SDK layer in exporters, or
 even out of process in Agent or Collector.
 
-API defines two interfaces - [`Sampler`](#sampler) and [`Decision`](#decision)
+OpenTelemetry API has two properties responsible for the data collection:
+
+* `IsRecordEvents` field of a `Span`
+  * If `true` the current `Span` records tracing events (attributes, events,
```
Comment (Oberon00) [inline specification/sdk-tracing.md]: I wouldn't make this a sublist of bullet points but rather just a single paragaph at the top-level point.

File: specification/sdk-tracing.md [hunk 16]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+    status, etc.), otherwise all tracing events are dropped.
+  * Users can use this property to determine if expensive trace events can be
+    avoided.
+  * [Span Processor](#span-processor) will receive all spans with this flag set.
+    However, [Span Exporter](#span-exporter) will not receive them unless the
+    `Sampled` flag was set.
+* `Sampled` `TraceFlag` on `SpanContext`
```
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
* `Sampled` flag in `TraceFlags` on `SpanContext`
```

File: specification/sdk-tracing.md [hunk 17]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+  * This flag is propagated via the `TraceFlag` to the child Spans. For more
```
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
  * This flag is propagated via the `SpanContext` to child Spans. For more
```

File: specification/sdk-tracing.md [hunk 18]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+    details see the w3c definition [here][trace-flags].
```
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
    details see the [W3C specification][trace-flags].
```

File: specification/sdk-tracing.md (hunk 7 above)
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
  * This flag indicates that the `Span` has been `sampled` and will be exported.
```

File: specification/sdk-tracing.md (hunk 12 above)
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
The flag combination `SampledFlag == false` and `IsRecording == true` means
```

File: specification/sdk-tracing.md (hunk 13 above)
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
The flag combination `SampledFlag == true` and `RecordEvents == false` could cause
```

File: specification/sdk-tracing.md [hunk 19]
```diff
@@ -23,7 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+
+The flag combination `SampledFlag == false` and `IsRecordEvents == true` means
+that the current `Span` does record information, but most likely the child
+`Span` will not.
+
+The flag combination `SampledFlag == true` and `RecordEvents == false` can cause
+gaps in the distributed trace, and because of this OpenTelemetry API MUST NOT
+allow this combination.
+
+SDK defines two interfaces - [`Sampler`](#sampler) and [`Decision`](#decision)
```
Comment (Oberon00) [inline specification/sdk-tracing.md]:
```suggestion
The SDK defines the two interfaces [`Sampler`](#sampler) and [`Decision`](#decision)
```

File: specification/sdk-tracing.md (hunk 17 above)
Comment (Oberon00) [inline specification/sdk-tracing.md]: Same as above: I wouldn't make this a sublist of bullet points but rather just a single paragaph at the top-level point.

File: specification/api-tracing.md [hunk 20]
```diff
@@ -287,12 +287,28 @@ The Span interface MUST provide:
...
+[SpanContext](#spancontext)).
+
+This flag may be `true` despite the entire trace being sampled out. This
+allows to record and process information about the individual Span without
+sending it to the backend. An example of this scenario may be recording and
+processing of all incoming requests for the processing and building of
+SLA/SLO latency charts while sending only a subset - sampled spans - to the
+backend. See also the [sampling section of SDK design](sdk-tracing.md#sampling).
+
+Users of the API should only access the `IsRecordingEvents` property when
```
Comment (bogdandrutu) [inline specification/api-tracing.md]:
```suggestion
Users of the API should only access the `IsRecording` property when
```

File: specification/sdk-tracing.md [hunk 21]
```diff
@@ -23,8 +23,32 @@ optimize the amount of telemetry that needs to be collected.
 All other sampling algorithms may be implemented on SDK layer in exporters, or
 even out of process in Agent or Collector.
 
-API defines two interfaces - [`Sampler`](#sampler) and [`Decision`](#decision)
-as well as a set of [built-in samplers](#built-in-samplers).
+The OpenTelemetry API has two properties responsible for the data collection:
+
+* `IsRecordingEvents` field of a `Span`. If `true` the current `Span` records
```
Comment (bogdandrutu) [inline specification/sdk-tracing.md]:
```suggestion
* `IsRecording` field of a `Span`. If `true` the current `Span` records
```

File: specification/sdk-tracing.md [hunk 22]
```diff
@@ -23,8 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+  all spans with this flag set. However, [Span Exporter](#span-exporter) will
+  not receive them unless the `Sampled` flag was set.
+* `Sampled` flag in `TraceFlags` on `SpanContext`. This flag is propagated via
+  the `SpanContext` to child Spans. For more details see the [W3C
+  specification][trace-flags]. This flag indicates that the `Span` has been
+  `sampled` and will be exported. [Span Processor](#span-processor) and [Span
+  Exporter](#span-exporter) will receive spans with the `Sampled` flag set for
+  processing.
+
+The flag combination `SampledFlag == false` and `IsRecordingEvents == true`
```
Comment (bogdandrutu) [inline specification/sdk-tracing.md]:
```suggestion
The flag combination `SampledFlag == false` and `IsRecording == true`
```

File: specification/sdk-tracing.md [hunk 23]
```diff
@@ -23,8 +23,32 @@ optimize the amount of telemetry that needs to be collected.
...
+means that the current `Span` does record information, but most likely the child
+`Span` will not.
+
+The flag combination `SampledFlag == true` and `IsRecordingEvents == false`
```
Comment (bogdandrutu) [inline specification/sdk-tracing.md]:
```suggestion
The flag combination `SampledFlag == true` and `IsRecording == false`
```

File: specification/api-tracing.md [hunk 24]
```diff
@@ -291,12 +291,28 @@ The Span interface MUST provide:
...
-the `AddEvent` operation and attributes using `SetAttributes`.
+This flag SHOULD be used to avoid expensive computations of a Span attributes or
+events in case when a Span is definitely not recorded. Note that any child
+span's recording is determined independently from the value of this flag
+(typically based on the `sampled` flag of a `TraceFlag` on
+[SpanContext](#spancontext)).
+
+This flag may be `true` despite the entire trace being sampled out. This
+allows to record and process information about the individual Span without
+sending it to the backend. An example of this scenario may be recording and
```
Comment (bogdandrutu) [inline specification/api-tracing.md]: These spans may also be sent to the backend. Usually they are not but they may be.

File: specification/api-tracing.md [hunk 25]
```diff
@@ -291,12 +291,28 @@ The Span interface MUST provide:
...
+processing of all incoming requests for the processing and building of
+SLA/SLO latency charts while sending only a subset - sampled spans - to the
+backend. See also the [sampling section of SDK design](sdk-tracing.md#sampling).
+
+Users of the API should only access the `IsRecording` property when
+instrumenting code and never access `SampledFlag` unless used in context
+propagators.
```
Comment (bogdandrutu) [inline specification/api-tracing.md]: Or Samplers

File: specification/api-tracing.md (hunk 24 above)
Comment (SergeyKanzhelev) [reply specification/api-tracing.md]: by processors?

File: specification/api-tracing.md (hunk 25 above)
Comment (SergeyKanzhelev) [reply specification/api-tracing.md]: users of API... Samplers are in SDK

**Comment by SergeyKanzhelev** [comment]
this change explains existing fields and does not introduce a change. It was active for 8 days and received the required number of approvals. Merging

File: specification/api-tracing.md (hunk 24 above)
Comment (bogdandrutu) [reply specification/api-tracing.md]: If some exporter wants to export this why not?

**Pull request merged**